### ISSUE TITLE
V2 MetaFunction doesn't infer loaderdata type properly with typescript

### ISSUE THREAD

**Issue by eavonius** [remix-run/remix#6612]
### What version of Remix are you using?

1.17.0

### Are all your remix dependencies & dev-dependencies using the same version?

- [X] Yes

### Steps to Reproduce

1. Create a route with typescript that has a loader function e.g.:

   ```typescript
   import type { V2_MetaFunction } from "@remix-run/node";
   import { useLoaderData } from "@remix-run/react";
   import { json, type LinksFunction, type LoaderFunction } from "@remix-run/server-runtime";
   
   type LoaderData = {
     someLoaderData: {x: string, y: number};
   }

   export const loader: LoaderFunction = async ({ params }) => {
     return json<LoaderData>({
       someLoaderData: {x: "One", y: 2}
     });
   };
   ````
   
2. Add a V2 meta function to access the data:

   ```typescript
   export const meta: V2_MetaFunction<typeof loader> = ({	data }) => {

     // First issue: This line works, but there is no type checking for someLoaderData it is "any"
     let x = data.someLoaderData.x;

     return [],
   };
   ```

3. Modify the V2 meta function to access the data:

   ```typescript
   export const meta: V2_MetaFunction<typeof loader> = ({	
     data 
   }: {
     data: LoaderData;
   }) => {

     // Second issue: This line now provides type checking properly, and works at runtime, but 
     // running a typecheck (or viewing the red underline error in vscode) shows the meta function 
     // as being declared with a data parameter that doesn't match the expected type.
     let x = data.someLoaderData.x;

     return [],
   };
   ```

### Expected Behavior

I would hope that the declaration in step 2 would infer the type properly and still provide strongly typed checking in vscode.

Failing that, the second declaration shouldn't generate an error when typechecking.

### Actual Behavior

The two issues described above in the comments.

**Comment by eavonius** [comment]
@herecydev hold on there, that doesn't fix the issue at all.

This makes the loader's args untyped, and now the params (and request) don't have types.

**Comment by MichaelDeBoey** [comment]
@eavonius Oh I missed not typing the args
The following snippet should work

```ts
import type { LoaderArgs, V2_MetaFunction } from "@remix-run/node";
import { json } from "@remix-run/node";
import { useLoaderData } from "@remix-run/react";

export const loader = async ({ params }: LoaderArgs) => {
  return json({ someLoaderData: {x: "One", y: 2} });
};
```

**Comment by eavonius** [comment]
> @eavonius Oh I missed not typing the args The following snippet should work
> 
> ```ts
> import type { LoaderArgs } from "@remix-run/node";
> import { json } from "@remix-run/node";
> 
> export const loader = async ({ params }: LoaderArgs) => {
>   return json({ someLoaderData: {x: "One", y: 2} });
> };
> ```

Yeah no, this doesn't provide the type inference for the data variable properly in the meta function in typescript.

**Comment by MichaelDeBoey** [comment]
@eavonius Then something must be wrong on your setup, as I do get `x` is of type `string | undefined` on a clean install
![](https://github.com/remix-run/remix/assets/6643991/01962377-645a-41a5-869a-d75fbbfb9518)

**Comment by eavonius** [comment]
> @eavonius Then something must be wrong on your setup, as I do get `x` is of type `string | undefined` on a clean install ![](https://user-images.githubusercontent.com/6643991/246605910-01962377-645a-41a5-869a-d75fbbfb9518.png)

Thanks @MichaelDeBoey this works! It only doesn't work when using the remix-typedjson package. I'll hop over there and see if they can resolve it, but for regular json serializations this does work properly. Thanks you and @herecydev for helping me figure out the proper combination of method signatures for this to work properly. Some of the existing examples show LoaderFunction being specified, which is probably what threw me off.

**Comment by MichaelDeBoey** [comment]
@eavonius Can you point me towards these examples you're talking about?
I'll see what I can do to get them updated

**Comment by eavonius** [comment]
> @eavonius Can you point me towards these examples you're talking about? I'll see what I can do to get them updated

Sure thing. I'll see if I can find what threw me off.

**Comment by eavonius** [comment]
> @eavonius Can you point me towards these examples you're talking about? I'll see what I can do to get them updated

Welp, I've looked everywhere and I can't find anything official that references LoaderFunction. So at this point, my guess is I either read it in a medium article or something talking about v2, or when I added the v2 MetaFunction, I inadvertently added that to my loader function calls too for consistency.

Looking forward to v2 coming out! Thanks for your help you two. Appreciate the follow up and helping me get this resolved.
